src/consensus/aft/raft.h

```diff
@@ -632,24 +673,65 @@ namespace aft
           {
           }
         }
+      }
+      catch (const std::logic_error& err)
+      {
+        LOG_FAIL_EXC(err.what());
+        return;
+      }
+
+      if (execute_async)
+      {
+        auto async_pending_msg =
+          std::make_unique<threading::Tmsg<AsyncPendingExec>>(
+            [](std::unique_ptr<threading::Tmsg<AsyncPendingExec>> msg) {
+              msg->data.pending_execution->async_execute();
+
+              msg->reset_cb(add_to_pending_execution_cb);
+              threading::ThreadMessaging::thread_messaging.add_task(
+                threading::MAIN_THREAD_ID, std::move(msg));
+            },
+            this,
+            std::move(aee));
 
-        if (!is_execution_pending)
+        if (threading::ThreadMessaging::thread_count > 1)
         {
-          aee->execute();
+          threading::ThreadMessaging::thread_messaging.add_task(
+            threading::ThreadMessaging::get_execution_thread(
+              ++next_exec_thread),
+            std::move(async_pending_msg));
         }
         else
         {
-          pending_executions.push_back(std::move(aee));
+          async_pending_msg->data.pending_execution->async_execute();
+          add_to_pending_execution_cb(std::move(async_pending_msg));
```

Comment:
I really can't parse this. Can you please add an explanatory comment?

We create a task which calls `async_execute`, and if we have multiple threads we send that task round-robin to another execution thread - I think I get that bit. But the task also calls `reset_cb`? And then sends itself to the main thread? And if we're single-threaded we execute inline and then add a message which is going to call `async_execute` again? If `execute_async` is true, I'd expect this to go through the messaging system (ie - be called later, asynchronously), even when single-threaded.

Reply:
Added a comment and broke this apart to make it easier to undestand